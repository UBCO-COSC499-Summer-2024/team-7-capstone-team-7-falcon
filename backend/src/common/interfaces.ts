import { PageOptionsDto } from '../dto/page-options.dto';

/**
 * Google OAuth user payload interface
 */
export interface OAuthGoogleUserPayload {
  email: string;
  given_name: string;
  family_name: string;
  picture: string;
}

/*
 * Jobs queue service interface
 */
export interface IJobQueueService {
  createJob(data: any): Promise<any>;
  getJobById(jobId: string): Promise<any>;
  pickUpJob(): Promise<any>;
  markJobAsComplete(jobId: string, result: any): Promise<void>;
}

/**
 * Bubble sheet payload interface
 */
export interface IBubbleSheetPayload {
  numberOfQuestions: number;
  defaultPointsPerQuestion: number;
  numberOfAnswers: number;
}

/**
 * Page meta dto parameters
 */
export interface PageMetaDtoParameters {
  pageOptionsDto: PageOptionsDto;
  itemCount: number;
}

/**
 * Individual exam interface for upcoming exams
 */
interface UpcomingExamsExamInterface {
  id: number;
  name: string;
  examDate: number;
}

/**
 * Upcoming exams interface
 */
export interface UpcomingExamsInterface {
  courseId: number;
  courseName: string;
  courseCode: string;
  exams: UpcomingExamsExamInterface[];
}

/**
 * Graded exam interface
 */
interface GradedSubmissionInterface {
  examId: number;
  examName: string;
  examDate: number;
  examReleasedAt: number;
  examScore: number;
  courseId: number;
}

/**
 * Graded exams interface
 */
export interface GradedSubmissionsInterface {
  exams: GradedSubmissionInterface[];
}

/**
 * User submission exam course details interface
 */
interface UserSubmissionExamCourseDetailsInterface {
  id: number;
  courseName: string;
  courseCode: string;
}

/**
 * User submission exam student submission interface
 */
interface UserSubmissionExamStudentSubmissionInterface {
  id: number;
  score: number;
  hasStudent?: boolean;
}

/**
 * User submission exam details interface
 */
interface UserSubmissionExamDetailsInterface {
  id: number;
  name: string;
  examDate: number;
}

/**
 * User submission exam interface
 */
export interface UserSubmissionExamInterface {
  exam: UserSubmissionExamDetailsInterface;
  course: UserSubmissionExamCourseDetailsInterface;
  grades?: number[];
  studentSubmission: UserSubmissionExamStudentSubmissionInterface;
  answers?: JSON;
}

/**
 * Course analytics submission student interface
 */
interface CourseAnalyticsSubmissionStudentInterface {
  student: {
    id: number;
    firstName: string;
    lastName: string;
    submissionScore: number;
    avatarUrl: string;
  };
}

/**
 * Course analytics submission interface
 */
interface CourseAnalyticsSubmissionInterface {
  submissions: CourseAnalyticsSubmissionStudentInterface[];
  exam: {
    id: number;
    title: string;
  };
}

/**
 * Course analytics response interface
 */
export interface CourseAnalyticsResponseInterface {
  courseMembersSize: number;
  courseExamsCount: number;
  examSubmissionsCount: number;
  examSubmissions: CourseAnalyticsSubmissionInterface[];
}

/**
 * User role count interface
 */
export interface UserRoleCount {
  count: number;
  role: string;
}

/**
 * Course details interface
 */
export interface CourseDetailsInterface {
  courseId: number;
  courseCode: string;
  semesterName: string;
  members: number;
  creator: {
    firstName: string;
    lastName: string;
  };
}

/**
 * Exam submissions with disputes count interface
 */
export interface ExamSubmissionsDisputesInterface {
  examId: number;
  examName: string;
  numberOfDisputes: number;
}
