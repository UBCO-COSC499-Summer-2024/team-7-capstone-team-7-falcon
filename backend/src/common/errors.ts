import { ERROR_MESSAGES } from '.';

/**
 * Exception to be thrown when a user already exists with a different auth type
 */
export class UserAlreadyExistsException extends Error {
  constructor() {
    super(ERROR_MESSAGES.authController.userAlreadyExists);
  }
}

/**
 * Exception to be thrown when an invalid auth method is provided
 */
export class InvalidAuthMethodException extends Error {
  constructor(message?: string) {
    super(message ?? ERROR_MESSAGES.authController.invalidAuthMethod);
  }
}

/**
 * Exception to be thrown when an error occurs authenticating with Google
 */
export class OAuthGoogleErrorException extends Error {
  constructor(message?: string) {
    super(message ?? ERROR_MESSAGES.authController.googleAuthError);
  }
}

/**
 * Exception to be thrown when a user is not found
 */
export class UserNotFoundException extends Error {
  constructor(message?: string) {
    super(message ?? ERROR_MESSAGES.userController.userNotFound);
  }
}

/**
 * Exception to be thrown when a user is not allowed to edit / create employee ID
 */
export class EmployeeIdAlreadyExistsException extends Error {
  constructor() {
    super(ERROR_MESSAGES.userController.employeeIdAlreadyExists);
  }
}

/**
 * Exception to be thrown when a user is not allowed to edit another user
 */
export class StudentIdAlreadyExistsException extends Error {
  constructor() {
    super(ERROR_MESSAGES.userController.studentIdAlreadyExists);
  }
}

/**
 * Exception to be thrown when a course is not found
 */
export class CourseNotFoundException extends Error {
  constructor(message?: string) {
    super(message ?? ERROR_MESSAGES.courseController.courseNotFound);
  }
}

/**
 * Exception to be thrown when an invalid invite code is provided
 */
export class InvalidInviteCodeException extends Error {
  constructor() {
    super(ERROR_MESSAGES.courseController.invalidInviteCode);
  }
}

/**
 * Exception to be thrown when an invalid invite code is provided
 */
export class UniqueConstraintException extends Error {
  constructor() {
    super(ERROR_MESSAGES.courseController.uniqueConstraintViolation);
  }
}

/**
 * Exception to be thrown when a course instance from an archived course is accessed
 */
export class CourseArchivedException extends Error {
  constructor() {
    super(ERROR_MESSAGES.courseController.courseArchived);
  }
}

/**
 * Exception to be thrown when a semester cannot be found
 */
export class SemesterNotFoundException extends Error {
  constructor() {
    super(ERROR_MESSAGES.semesterController.semesterNotFound);
  }
}

/**
 * Exception to be thrown when semester is not possible to create
 */
export class SemesterCreationException extends Error {
  constructor(message: string) {
    super(message);
  }
}

/**
 * Exception to be thrown when a job creation fails
 */
export class JobCreationException extends Error {
  constructor(message: string) {
    super(message);
  }
}

/**
 * Exception to be thrown when a job could not be completed
 */
export class CouldNotCompleteJobException extends Error {
  constructor(message: string) {
    super(message);
  }
}

/**
 * Exception to be thrown when a job is not found
 */
export class JobNotFoundException extends Error {
  constructor() {
    super(ERROR_MESSAGES.queueController.jobNotFound);
  }
}

/**
 * Exception to be thrown when exam was failed to create
 */
export class ExamCreationException extends Error {
  constructor(message: string) {
    super(message);
  }
}

/**
 * Exception thrown when none of the student and employee fields are provided
 */
export class UserStudentEmployeeFieldException extends Error {
  constructor() {
    super(
      ERROR_MESSAGES.userController.userStudentEmployeeIdFieldsMissingError,
    );
  }
}

/**
 * Exception thrown when token is invalid or not found
 */
export class TokenInvalidException extends Error {
  constructor() {
    super(ERROR_MESSAGES.tokenController.invalidToken);
  }
}

/**
 * Exception thrown when token is expired
 */
export class TokenExpiredException extends Error {
  constructor() {
    super(ERROR_MESSAGES.tokenController.tokenExpired);
  }
}

/**
 * Exception thrown when email is not verified
 */
export class EmailNotVerifiedException extends Error {
  constructor() {
    super(ERROR_MESSAGES.authController.emailNotVerified);
  }
}

/**
 * Exception thrown when password is invalid
 */
export class InvalidPasswordException extends Error {
  constructor() {
    super(ERROR_MESSAGES.authController.invalidPassword);
  }
}

/**
 * Exception thrown when exam is not found
 */
export class ExamNotFoundException extends Error {
  constructor() {
    super(ERROR_MESSAGES.examController.examNotFound);
  }
}

/**
 * Exception thrown when user submission is not found
 */
export class UserSubmissionNotFound extends Error {
  constructor() {
    super(ERROR_MESSAGES.examController.userSubmissionNotFound);
  }
}

/**
 * Exception thrown when submission is not found
 */
export class SubmissionNotFoundException extends Error {
  constructor() {
    super(ERROR_MESSAGES.examController.submissionNotFound);
  }
}

/**
 * Exception thrown when file is not found
 */
export class FileNotFoundException extends Error {
  constructor() {
    super(ERROR_MESSAGES.common.fileNotFound);
  }
}

/**
 * Exception thrown when there is an error with course role
 */
export class CourseRoleException extends Error {
  constructor(message: string) {
    super(message);
  }
}
