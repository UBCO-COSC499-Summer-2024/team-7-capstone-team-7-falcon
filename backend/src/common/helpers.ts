import { Request } from 'express';

/**
 * Get a cookie from the request
 * @param req {Request} - The request object
 * @param cookieName {string} - The name of the cookie to get
 * @returns {string} - The cookie value or null
 */
export function getCookie(req: Request, cookieName: string): string {
  const cookieHeader = req.headers?.cookie ?? req.body?.headers?.cookie;
  console.log('req.cookies', req.cookies);
  if (!cookieHeader) {
    return null;
  }

  const cookies: string[] = cookieHeader.split(';');
  for (const cookie of cookies) {
    const [name, value] = cookie.split('=');
    if (name.trim() === cookieName) {
      return value;
    }
  }
}
