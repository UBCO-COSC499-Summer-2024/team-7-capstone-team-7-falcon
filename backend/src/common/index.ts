export const ERROR_MESSAGES = {
  common: {
    passwordString: 'Password must be a string',
    passwordRequired: 'Password is required',
    confirmPasswordRequired: 'Confirmation password is required',
    confirmPasswordString: 'Confirmation password must be a string',
    passwordStrong:
      'Password must be at least 8 characters long and contain at least one lowercase letter, one uppercase ' +
      'letter, one number, and one symbol',
    passwordMustMatch: 'Password and confirmation password do not match',
    emailRequired: 'Email is required',
    emailInvalid: 'Email is invalid',
  },
  userController: {
    userNotFound: 'User not found',
    editForbidden: 'You are not allowed to edit this user',
    employeeIdAlreadyExists: 'Employee ID already exists',
    studentIdAlreadyExists: 'Student ID already exists',
    userStudentEmployeeIdFieldsMissingError:
      'Student or employee ID fields are missing',
  },
  authController: {
    userAlreadyExists: 'User already exists with different auth type',
    invalidAuthMethod: 'Invalid auth method',
    googleAuthError: 'Error authenticating with Google',
    emailNotVerified: 'Email not verified',
    notSupportedAuthType: 'User account has unsupported authentication type',
    invalidPassword: 'Invalid password',
    studentIdMissing: 'Student ID is missing',
  },
  tokenController: {
    invalidToken: 'Invalid token',
    tokenExpired: 'Token has expired',
    tokenRequired: 'Token is required',
    tokenString: 'Token must be a string',
  },
  userEditDto: {
    firstNameString: 'First name must be a string',
    firstNameRequired: 'First name is required',
    firstNameLength: 'First name must be between 2 and 15 characters',
    studentNumberMustBeNumber: 'Student number must be a number',
    studentNumberMustBeGreater: 'Student number must be greater than 0',
    employeeNumberMustBeNumber: 'Employee number must be a number',
    employeeNumberMustBeGreater: 'Employee number must be greater than 0',
  },
  userCreateDto: {
    emailRequired: 'Email is required',
    passwordString: 'Password must be a string',
    passwordRequired: 'Password is required',
    confirmPasswordRequired: 'Confirmation password is required',
    confirmPasswordString: 'Confirmation password must be a string',
    passwordStrong:
      'Password must be at least 8 characters long and contain at least one lowercase letter, one uppercase letter, one number, and one symbol',
    passwordMustMatch: 'Password and confirmation password do not match',
  },
  courseController: {
    courseNotFound: 'Course not found',
    invalidInviteCode: 'Invalid invite code',
    uniqueConstraintViolation: 'Unique constraint violated',
    courseCreationFailed: 'Course could not be created',
    coursesNotFound: 'Courses could not be found',
    courseArchived: 'Course is archived. Cannot be accessed.',
    userNotEnrolledInCourse: 'User is not enrolled in the course',
  },
  courseEnrollDto: {
    inviteCodeRequired: 'Invite code is required',
  },
  semesterCreateDto: {
    nameRequired: 'Name is required',
    nameLength: 'Name must be between 2 and 15 characters',
    startsAtRequired: 'Start date is required',
    endsAtRequired: 'End date is required',
  },
  semesterController: {
    semesterStartDateMustBeBeforeEndDate: 'Start date must be before end date',
    semesterStartDateMustBeTwoDaysAhead:
      'Start date must be at least two days ahead',
    semesterNotFound: 'Semester not found',
    semestersNotFound: 'Semesters not found',
  },
  queueController: {
    jobNotFound: 'Job not found',
  },
  examController: {
    examDateError: 'Exam date must be in the future',
    noUpcomingExamsFound: 'No upcoming exams found',
    examNotFound: 'Exam not found',
  },
};
