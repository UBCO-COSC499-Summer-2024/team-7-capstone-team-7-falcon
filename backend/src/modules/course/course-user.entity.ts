import { CourseRoleEnum } from '../../enums/user.enum';
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { UserModel } from '../user/entities/user.entity';
import { CourseModel } from './course.entity';

@Entity('course_user_model')
export class CourseUser extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ default: CourseRoleEnum.STUDENT })
  course_role: CourseRoleEnum;

  @ManyToOne(() => UserModel, (user) => user.courses)
  @JoinColumn({ name: 'user_id' })
  user: UserModel;

  @ManyToOne(() => CourseModel, (course) => course.users)
  @JoinColumn({ name: 'course_id' })
  course: CourseModel;
}
