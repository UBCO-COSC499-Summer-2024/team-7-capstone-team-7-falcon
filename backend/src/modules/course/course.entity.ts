import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  BaseEntity,
  OneToMany,
} from 'typeorm';
import { CourseUser } from './course-user.entity';
import { ExamModel } from '../exams/entities/exam.entity';

@Entity('course_model')
export class CourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  course_code: string;

  @Column()
  course_name: string;

  @Column({ type: 'bigint' })
  created_at: number;

  @Column({ type: 'bigint' })
  updated_at: number;

  // add term id later

  @Column({ default: false })
  is_archived: boolean;

  @Column()
  invite_code: string;

  @Column()
  section_name: string;

  @OneToMany(() => CourseUser, (user) => user.course)
  users: CourseUser[];

  @OneToMany(() => ExamModel, (exam) => exam.course)
  exams: ExamModel[];
}
