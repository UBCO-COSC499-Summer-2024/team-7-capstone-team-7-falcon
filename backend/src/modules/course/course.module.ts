import { Module } from '@nestjs/common';
import { CourseController } from './course.controller';
import { CourseService } from './course.service';
import { UserService } from '../user/user.service';

@Module({
  controllers: [CourseController],
  providers: [CourseService, UserService],
  exports: [CourseService],
})
export class CourseModule {}
