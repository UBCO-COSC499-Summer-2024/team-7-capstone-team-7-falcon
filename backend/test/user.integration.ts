import { HttpStatus } from '@nestjs/common';
import { UserModel } from '../src/modules/user/entities/user.entity';
import { UserModule } from '../src/modules/user/user.module';
import { setUpIntegrationTests, signJwtToken } from './utils/testUtils';
import { CourseRoleEnum, UserRoleEnum } from '../src/enums/user.enum';
import { EmployeeUserModel } from '../src/modules/user/entities/employee-user.entity';
import { StudentUserModel } from '../src/modules/user/entities/student-user.entity';
import { CourseModel } from '../src/modules/course/entities/course.entity';
import { CourseUserModel } from '../src/modules/course/entities/course-user.entity';

describe('User Integration', () => {
  const supertest = setUpIntegrationTests(UserModule);

  beforeEach(async () => {
    await CourseUserModel.delete({});
    await CourseModel.delete({});
    await UserModel.delete({});

    await CourseModel.query(
      'ALTER SEQUENCE course_model_id_seq RESTART WITH 1',
    );
    await UserModel.query('ALTER SEQUENCE user_model_id_seq RESTART WITH 1');
    await CourseUserModel.query(
      'ALTER SEQUENCE course_user_model_id_seq RESTART WITH 1',
    );
  });

  describe('GET /', () => {
    it('should return status 401 when no token is provided', () => {
      return supertest().get('/user').expect(HttpStatus.UNAUTHORIZED);
    });

    it('should return status 200 and user object', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const response = await supertest()
        .get('/user')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`]);

      expect(response.status).toBe(HttpStatus.OK);
    });
  });

  describe('GET /:uid', () => {
    it('should return status 401 when no token is provided', () => {
      return supertest().get('/user/1').expect(HttpStatus.UNAUTHORIZED);
    });

    it('should return status 401 when user is not an admin', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      return supertest()
        .get('/user/1')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(HttpStatus.UNAUTHORIZED);
    });

    it('should return status 400 when uid is not a number', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      return supertest()
        .get('/user/abc')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(HttpStatus.BAD_REQUEST)
        .expect({
          message: 'Validation failed (numeric string is expected)',
          error: 'Bad Request',
          statusCode: HttpStatus.BAD_REQUEST,
        });
    });

    it('should return status 200 and user object', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      const response = await supertest()
        .get(`/user/${user.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(1)}`]);

      expect(response.status).toBe(HttpStatus.OK);
      expect(response.body).toMatchSnapshot();
    });

    it("should return status 404 when user id doesn't exist", async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      return supertest()
        .get('/user/100')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(HttpStatus.NOT_FOUND)
        .expect({
          message: 'User not found',
        });
    });
  });

  describe('PATCH /:uid', () => {
    it('should return status 401 when no token is provided', () => {
      return supertest().patch('/user/1').expect(HttpStatus.UNAUTHORIZED);
    });

    it('should return status 400 when uid is not a number', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      return supertest()
        .patch('/user/abc')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .expect(HttpStatus.BAD_REQUEST)
        .expect({
          message: 'Validation failed (numeric string is expected)',
          error: 'Bad Request',
          statusCode: HttpStatus.BAD_REQUEST,
        });
    });

    it('should return status 400 when first_name is not provided', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      return supertest()
        .patch(`/user/${user.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({})
        .expect(HttpStatus.BAD_REQUEST)
        .expect({
          message: [
            'First name must be between 2 and 15 characters',
            'First name is required',
            'First name must be a string',
          ],
          error: 'Bad Request',
          statusCode: HttpStatus.BAD_REQUEST,
        });
    });

    it('should return 403 when user is not allowed to edit another user', async () => {
      const userOne = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const userTwo = await UserModel.create({
        first_name: 'Jane',
        last_name: 'Doe',
        email: 'jane.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      return supertest()
        .patch(`/user/${userTwo.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(userOne.id)}`])
        .send({ first_name: 'John' })
        .expect(HttpStatus.FORBIDDEN)
        .expect({
          message: 'You are not allowed to edit this user',
        });
    });

    it('should return status 200 when user updated', async () => {
      const user = await UserModel.create({
        first_name: 'Jane',
        last_name: 'Doe',
        email: 'jane.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const response = await supertest()
        .patch(`/user/${user.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({ first_name: 'John' });

      expect(response.status).toBe(HttpStatus.OK);
    });

    it('should return status 200 when user is admin and updates another user', async () => {
      const userOne = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      const userTwo = await UserModel.create({
        first_name: 'Jane',
        last_name: 'Doe',
        email: 'jane.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      return supertest()
        .patch(`/user/${userTwo.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(userOne.id)}`])
        .send({ first_name: 'John' })
        .expect(HttpStatus.OK)
        .expect({
          message: 'ok',
        });
    });

    it('should return status 404 when user id does not exist', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        role: UserRoleEnum.ADMIN,
      }).save();

      return supertest()
        .patch('/user/100')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({ first_name: 'John' })
        .expect(HttpStatus.NOT_FOUND)
        .expect({
          message: 'User not found',
        });
    });

    it('should return status 409 when employee_id already exists', async () => {
      const userOne = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const userTwo = await UserModel.create({
        first_name: 'Jane',
        last_name: 'Doe',
        email: 'jane.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const employeeIdRecord = await EmployeeUserModel.create({
        employee_id: 123,
        user: userTwo,
      }).save();

      userTwo.employee_user = employeeIdRecord;
      await userTwo.save();

      return supertest()
        .patch(`/user/${userOne.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(userOne.id)}`])
        .send({ first_name: 'John', employee_id: 123 })
        .expect(HttpStatus.CONFLICT)
        .expect({
          message: 'Employee ID already exists',
        });
    });

    it('should return status 409 when student_id already exists', async () => {
      const userOne = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const userTwo = await UserModel.create({
        first_name: 'Jane',
        last_name: 'Doe',
        email: 'jane.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const studentIdRecord = await StudentUserModel.create({
        student_id: 123,
        user: userTwo,
      }).save();

      userTwo.student_user = studentIdRecord;
      await userTwo.save();

      return supertest()
        .patch(`/user/${userOne.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(userOne.id)}`])
        .send({ first_name: 'John', student_id: 123 })
        .expect(HttpStatus.CONFLICT)
        .expect({
          message: 'Student ID already exists',
        });
    });

    it('should return status 200 when user updated with employee_id', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      return supertest()
        .patch(`/user/${user.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({ first_name: 'Jane', employee_id: 1, student_id: 1 })
        .expect(HttpStatus.OK)
        .expect({
          message: 'ok',
        });
    });

    it('should return status 500 when an error occurs', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      // Relies on the fact that previous test case creates a student_id record
      return supertest()
        .patch(`/user/${user.id}`)
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`])
        .send({ first_name: 'Jane', employee_id: 1, student_id: 1 })
        .expect(HttpStatus.INTERNAL_SERVER_ERROR);
    });
  });

  describe('GET /courses', () => {
    it('should return status 401 when no token is provided', () => {
      return supertest().get('/user/courses').expect(HttpStatus.UNAUTHORIZED);
    });

    it('should return status 404 when there are no courses', async () => {
      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      const response = await supertest()
        .get('/user/courses')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`]);

      expect(response.status).toBe(HttpStatus.NOT_FOUND);
    });

    it('should return status 200 and courses', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
      }).save();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await CourseUserModel.create({
        course_role: CourseRoleEnum.STUDENT,
        user: user,
        course: course,
      }).save();

      const response = await supertest()
        .get('/user/courses')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`]);

      expect(response.status).toBe(HttpStatus.OK);
      expect(response.body).toMatchSnapshot();
    });

    it('should return status 400 when user is enrolled in the course, but course is archived', async () => {
      const course = await CourseModel.create({
        course_code: 'COSC 499',
        course_name: 'Capstone Project',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
        section_name: '001',
        invite_code: '123',
        is_archived: true,
      }).save();

      const user = await UserModel.create({
        first_name: 'John',
        last_name: 'Doe',
        email: 'john.doe@test.com',
        password: 'password',
        created_at: 1_000_000_000,
        updated_at: 1_000_000_000,
      }).save();

      await CourseUserModel.create({
        course_role: CourseRoleEnum.STUDENT,
        user: user,
        course: course,
      }).save();

      const response = await supertest()
        .get('/user/courses')
        .set('Cookie', [`auth_token=${signJwtToken(user.id)}`]);

      expect(response.status).toBe(HttpStatus.NOT_FOUND);
    });
  });
});
