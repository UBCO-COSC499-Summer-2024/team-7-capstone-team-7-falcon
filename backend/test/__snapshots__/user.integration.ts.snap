// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Integration GET /:uid should return status 200 and user object 1`] = `
{
  "auth_type": "email",
  "avatar_url": null,
  "created_at": "1000000000",
  "email": "john.doe@test.com",
  "employee_user": null,
  "first_name": "John",
  "id": 1,
  "last_name": "Doe",
  "role": "admin",
  "student_user": null,
  "updated_at": "1000000000",
}
`;

exports[`User Integration GET /courses should return status 200 and courses 1`] = `
[
  {
    "course": {
      "course_code": "COSC 499",
      "course_name": "Capstone Project",
      "created_at": "1000000000",
      "id": 1,
      "invite_code": "123",
      "is_archived": false,
      "section_name": "001",
      "updated_at": "1000000000",
    },
    "course_role": "student",
    "id": 1,
    "user": {
      "auth_type": "email",
      "avatar_url": null,
      "created_at": "1000000000",
      "email": "john.doe@test.com",
      "first_name": "John",
      "id": 1,
      "last_name": "Doe",
      "password": "password",
      "role": "student",
      "updated_at": "1000000000",
    },
  },
]
`;
